File: 2023-puzzles/challenge-3/solutions/ruby/solution-1.rb
require 'json'

puzzle_data = JSON.parse(File.read('../../puzzle.json'))
result_data = JSON.parse(File.read('../../result.json'))

linked_list = puzzle_data['linkedList']
top = puzzle_data['top']

top_node = linked_list.find { |node| node['id'] == top }

tree = []

def traverse(node, tree, linked_list)
    return if node.nil?

    tree.push(node['value'])
    traverse(linked_list.find { |n| n['id'] == node['next'] }, tree, linked_list)
end

traverse(top_node, tree, linked_list)

puts tree.to_s
puts 'Is solution correct?', JSON.generate(tree) == JSON.generate(result_data)

File: 2023-puzzles/challenge-3/solutions/javascript/solution-1.js
const { linkedList, top } = require('../../puzzle.json');
const result = require('../../result.json');

const isSolutionCorrect = () => {
    return JSON.stringify(tree) === JSON.stringify(result);
};

const topNode = linkedList.find(node => node.id === top);
const tree = [];
const traverse = node => {
    if (!node) return;
    tree.push(node.value);
    traverse(linkedList.find(n => n.id === node.next));
    return tree;
};

traverse(topNode);

console.log(tree);
console.log('Is solution correct?', isSolutionCorrect());
